### ISSUE TITLE
Multiple Fonts

### ISSUE THREAD

**Issue by AndreKoepke** [vandeseer/easytable#34]
For a project we have different fonts. The regular fonts, for "normal" characters. And "fallback"-fonts, for ❤ (emojis).

Our example for this fallback-method looks like this: 
```JAVA
    private void printWithFallback(final String line, final int fontSize, final float startX, final float startY) throws IOException {
        this.contentStream.beginText();
        this.contentStream.newLineAtOffset(startX, startY);


        for (int offset = 0; offset < line.length(); ) {

            PDFont fontToPrint = this.document.getFonts().get(0);
            final int codePoint = line.codePointAt(offset);
            offset += Character.charCount(codePoint);
            String codePointAsString = new String(new int[]{codePoint}, 0, 1);

            //try all available fonts
            for (final PDFont fontToTry : this.document.getFonts()) {
                try {
                    fontToTry.encode(codePointAsString);
                    fontToPrint = fontToTry;
                    break;
                } catch (final IllegalArgumentException e) {
                    //try next font
                }
            }

            try {
                fontToPrint.encode(codePointAsString);
            } catch (final IllegalArgumentException e) {
                log.error("Codepoint {} could not be printed to PDF and has been replaced by the fallback character {}. " +
                                "This probably means that a font which can encode this codepoint should be included in the next release.",
                        codePoint, FALLBACK_CHARACTER);
                codePointAsString = FALLBACK_CHARACTER;
            }

            this.contentStream.setFont(fontToPrint, fontSize);
            this.contentStream.showText(codePointAsString);
        }

        this.contentStream.endText();
    }
```

**Comment by vandeseer** [comment]
I think this is pretty closely related to https://github.com/vandeseer/easytable/issues/29. Because if the `TableDrawer` class would allow for easier reuse from outside one could just override the respective methods for this case.

**Comment by AndreKoepke** [comment]
That would be nice.

**Comment by vandeseer** [comment]
Please check out the `develop` branch, I updated the code accordingly and also added an example class (together with a hint in the `README`). For your specific use case it should exactly work as in the example class: [EasytableCustomTableDrawer](https://github.com/vandeseer/easytable/blob/develop/src/test/java/org/vandeseer/custom/EasytableCustomTableDrawer.java)

For installing the current snapshot version locally just use: `mvn install -DskipTests -Dgpg.skip` 
I didn't yet release this though – but may do so soon. 

If it works and you have some nice code and tables to share, please let me know. ;-)

**Comment by vandeseer** [comment]
Did it work @AndreKoepke? Just wanted to confirm it also from your side before I do the next release.

**Comment by AndreKoepke** [comment]
I didn't tested it yet. I will try to do it in the next few days.

**Comment by AndreKoepke** [comment]
https://github.com/vandeseer/easytable/blob/799929001c5f3184b2ba16fe6078d96972cc41c9/src/main/java/org/vandeseer/easytable/TableDrawer.java#L49

The constructor must be `public`. I cant use in the current version.

**Comment by AndreKoepke** [comment]
See https://github.com/vandeseer/easytable/pull/35

**Comment by vandeseer** [comment]
It is `protected` in the `develop` branch, as I pointed out above (see [here](https://github.com/vandeseer/easytable/blob/develop/src/main/java/org/vandeseer/easytable/TableDrawer.java#L49)).

You would need to install that branch manually via `mvn install -DskipTests -Dgpg.skip`. Then you can reuse the `TableDrawer` like so: 

    @SuperBuilder
    private static class CustomTableDrawer extends TableDrawer {

        CustomTableDrawer(float startX, float startY, PDPageContentStream contentStream, Table table, float endY) {
            super(startX, startY, contentStream, table, endY);
        }

        @Override
        protected void drawText(String text, PDFont font, int fontSize, Color color, float x, float y) throws IOException {
            System.out.println("My custom table drawer is called :-)");
            super.drawText(text.toUpperCase(), font, fontSize, color, x, y);
        }

    }

In case you are using lombok at least. I created two examples hence:
- [CustomTableDrawer.java](https://github.com/vandeseer/easytable/blob/develop/src/test/java/org/vandeseer/custom/EasytableCustomTableDrawer.java)
- [NoLombokCustomTableDrawer.java](https://github.com/vandeseer/easytable/blob/develop/src/test/java/org/vandeseer/custom/EasytableNoLombokCustomTableDrawer.java)

**Comment by AndreKoepke** [comment]
Whuups, was on the wrong branch. My bad. :D

**Comment by AndreKoepke** [comment]
Iam done. All our features working now, but easyTable needs some changes (look at pull-request).

```java
import lombok.Getter;
import lombok.experimental.SuperBuilder;
import org.apache.pdfbox.pdmodel.PDPageContentStream;
import org.apache.pdfbox.pdmodel.font.PDFont;
import org.vandeseer.easytable.TableDrawer;
import org.vandeseer.easytable.structure.Table;

import java.awt.*;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

@SuperBuilder
public class EasyTableDrawer extends TableDrawer {

    private static final String FALLBACK_CHARACTER = "?";
    private final List<PDFont> fonts;

    @Getter
    private float drawSizeCurrentPage;


    EasyTableDrawer(final float startX,
                    final float startY,
                    final PDPageContentStream contentStream,
                    final Table table,
                    final float endY,
                    final List<PDFont> fonts) {
        
        super(startX, startY, contentStream, table, endY);
        this.fonts = fonts;
    }

    @Override
    protected void drawText(final String text, final PDFont font, final int fontSize, final Color color, final float x, final float y) throws IOException {
        super.contentStream.beginText();
        super.contentStream.newLineAtOffset(x, y);


        for (int offset = 0; offset < text.length(); ) {

            PDFont fontToPrint = font;
            final int codePoint = text.codePointAt(offset);
            offset += Character.charCount(codePoint);
            String codePointAsString = new String(new int[]{codePoint}, 0, 1);

            final List<PDFont> fonts = new ArrayList<>(this.fonts);
            fonts.add(0, font);

            //try all available fonts
            for (final PDFont fontToTry : fonts) {
                try {
                    fontToTry.encode(codePointAsString);
                    fontToPrint = fontToTry;
                    break;
                } catch (final IllegalArgumentException e) {
                    //try next font
                }
            }

            try {
                fontToPrint.encode(codePointAsString);
            } catch (final IllegalArgumentException e) {
                codePointAsString = FALLBACK_CHARACTER;
            }

            super.contentStream.setFont(fontToPrint, fontSize);
            super.contentStream.showText(codePointAsString);
        }

        super.contentStream.endText();
    }

    @Override
    public void draw() throws IOException {
        final int startIndex = this.rowToDraw;

        super.draw();

        final int lastIndex;

        if (this.rowToDraw == startIndex) {
            // all draw
            lastIndex = this.table.getRows().size();
        } else {
            lastIndex = this.rowToDraw;
        }


        this.drawSizeCurrentPage = IntStream.range(startIndex, lastIndex)
                .mapToObj(index -> this.table.getRows().get(index).getHeight())
                .reduce(0f, (left, right) -> left + right);
    }


}
```

**Comment by vandeseer** [comment]
Glad you made it now and thanks for your pull request. I will look into it shortly and will then release a new version.

**Comment by vandeseer** [comment]
Included in version 0.4.2.
